enterprisemediawiki/meza
# Install apache httpd via yum, setup elasticsearch management, other fixes

This change has many changes rolled into it, which was not my hope. The main change is installing Apache httpd via yum, but several other changes are listed below. **This PR is being merged into the REL0_9 branch, not master.** I need a development branch to merge into so I can make smaller PRs. Subsequent PRs will merge smaller change-sets into dev branches, and then we can do major testing against release branches prior to merging them into master.

### Testing

* `curl -LO https://raw.githubusercontent.com/enterprisemediawiki/meza/yum-httpd/scripts/install.sh`
* `sudo bash install.sh`
* Use `yum-httpd` branch
* Perform normal tests.
* Add an authenticated user to access Elasticsearch by doing `sudo htpasswd -c /etc/httpd/.htpasswd <username>`. Enter a STRONG password. If you create a second user don't include the `-c` option.
* Access "kopf" elasticsearch plugin available at `http://<domain>:8008/_plugin/kopf`. See also `_plugin/elasticsearch-head`, `_plugin/bigdesk` and `_plugin/inquisitor`

### Changes

* Apache httpd is now installed via yum and is no longer compiled manually.
  * A lot of complexity was stripped from `apache.sh`
  * `httpd` is now installed in `yums.sh`
  * The "apache" user is now part of the group "apache" instead of "www" since that was the default.
  * Edited httpd.conf to fit with yum-installed version. Note that all the modules are not loaded directly in that file anymore, but by generally including all config files from within `conf.modules.d`. We can employ this method to allow customizations to apache config.
  * Several whitespace changes to `httpd.conf` to standardize that file. However, we're not standard across the project. We should add "tabs or X spaces" to CONTRIBUTING.md.
  * Several changes to VirtualHosts to support other services. See below.
  * Logrotate script modified to actually work
  * php.ini updated to point to non-thread-safe version of Opcache since that's the version of Apache and PHP's apache module that came with yum. From what I can tell this doesn't really matter. We may be able to squeak some more performance out of using a different php module, but it's not critical now.
  * php.ini also no longer has dev and ops option: just went with ops. They were basically the same anyway.
  * Removed FPM from PHP compilation since we're not using that as our httpd-php handler/module/whatever.
  * Installed mod_ssl and mod_proxy_html via yum to support HTTPS and proxying
* Change special Apache port for Parsoid to access MediaWiki's PHP API from 8142 to 9000
  * This is due to the fact that 8142 is not considered a standard port for Apache according to SELinux. By running `semanage port -l | egrep '(^http_port_t|6379)'` you can see that the standard ports are 80, 81, 443, 488, 8008, 8009, 8443, and 9000. 9000 was chosen because it seemed easiest to remember. Note that 8008 was chosen for the Elasticsearch reverse proxy for the same reasons. Additionally note that SELinux has been taken to "permissive" mode instead of "enforcing" (for reasons mentioned below), which means that it is not strictly required that these port standards be followed. Nothing is actually preventing the use of port 8142, but if SELinux were to be set back to "enforcing" it would be much easier to stay compliant with its restrictions.
  * SELinux was able to be made to work with Elasticsearch and all other aspects of meza, but in this configuration Parsoid ran very, very slowly. Saving a page when using Visual Editor took 30 seconds or more. From what I could tell many other web applications also want SELinux disabled or set to "permissive", but we can address putting it back in "enforcing" later. See issue #314.
  * Also removed comment in LocalSettings.php about using port 8142 for Parsoid if used the Parsoid Debian package because (1) we're not on Debian and (2) this seemed like it might be confusing considering all the port numbers floating around.
* Elasticsearch
  * Port 8008 opened in firewall in `apache.sh`
  * Created reverse-proxy in httpd.conf to allow GET-only access (no POST, PUT or DELETE) to Elasticsearch from outside the server. Access to this can be acheived via `http://<domain>:8008, but requires authentication. `install.sh` does not automatically setup a user to get through the authentication. You need to run `sudo htpasswd -c /etc/httpd/.htpasswd <username>` and then choose a STRONG password. 
  * Install plugin "kopf" available at `http://<domain>:8008/_plugin/kopf`
  * Install plugin "elasticsearch-head" available at `http://<domain>:8008/_plugin/elasticsearch-head`
  * Install plugin "bigdesk" available at `http://<domain>:8008/_plugin/TBD`
  * Install plugin "inquisitor" available at `http://<domain>:8008/_plugin/TBD`
  * Created issue #316 to down-select to two plugins at some point
* [WikiBlender](https://github.com/jamesmontalvo3/WikiBlender) used for wiki landing page
  * .gitignore updated so meza won't mess with this separate git repo
  * .htaccess update so apache will allow access to this new htdocs directory
  * /opt/meza/htdocs/index.php updated to just be effectively a redirect to WikiBlender
  * Added default `BlenderSettings.php` file. This will have to be considered when restructuring config files (issues #242 and #303)
* Added `commandTimes.js` which parses the "command log" (`/opt/meza/logs/*cmd.log`) and prints the time required for each script during install. This command is run at the end of `install.sh` to print out the totals for th user.
* Installed Vim via yum because syntax highlighting
* Installed net-tools via yum for `netstat` command

### Issues addressed

* Closes #312 (install apache httpd with yum)
* Closes #300 (Improve landing page)
* Closes #281 (install kopf)
* Closes #255 (install ES management tools)
* Closes #175 (httpd logrotate fail)
* Closes #138 (apache.sh several "to-do" statements)
* Closes #306 (Install Vim)
* Closes #135 (Apache missing SSL/EVP support)
* Closes #134 (Apache compiled without LDAP)
* Closes #267 (issues with Digital Ocean are not showing up anymore)
